criu/cr-check.c

```diff
@@ -516,6 +516,13 @@ static int check_ipc(void)
 {
 	int ret;
 
+	/*
+	 * There seems to be no capability for accessing sem_next_id.
+	 * Let's check if root or not.
+	 */
+	if (opts.uid)
```

Comment:
sem_next_id is protected by CAP_CHECKPOINT_RESTORE, isn't it?

Reply:
Yes, since kernel 5.16 (rc1) that is the case. Since this bit is part of `criu check` should we remove it explicitly or only for kernels 5.16 and later?

Reply:
I've changed this to check the kernel version in unprivileged mode. On kernel >= 5.16 we'll check that `sem_next_id` is accessible. On older kernels we'll skip it, since there's nothing the user can do besides run as root.

Reply:
I would prefer if you could avoid using version checks. For kernels using backporting the version check will not work.

Reply:
True. I've reverted this back to skipping the check for non-root because `access()` runs with an empty set of caps and will always fail for non-root users. I didn't realize that before.

Checking `opts.unprivileged` is pointless because 1) non-root users must have it set, otherwise we don't reach this code and 2) if a root uses `--unprivileged` for whatever reason we'll skip the check but we don't skip actually trying to restore `sem_next_id` if it was saved, so we'd just lose an opportunity to catch a problem.

Reply:
I would argue that this logic is somewhat faulty. `sem_next_id` is not always restored, yes, but I might be fine with that and rather see a warning than an error that prevents other checks from running. It's not intuitive either: it feels weird that a check passes for a non-root user and fails in a more privileged context, rather than the other way round. It took me a while to figure out why that happened, I don't think I'd realise the reason without consulting the source code.

Reply:
Just so I'm clear, you're not concerned about the new behaviour that this PR introduced (which is _`check_ipc()` always returns success for non-root users_), right? Rather, you are suggesting that under root this check should be non-fatal so the later checks can be performed?

If this is the case can you open a new issue referencing this comment and make your suggestion there? This check was added in https://github.com/checkpoint-restore/criu/commit/03791bc2c3059d2412f4c19ec4b9b71540792a62 when all checks were fatal. Non-fatal checks were added in https://github.com/checkpoint-restore/criu/commit/70cb4326b61fde36edf669e536fddc3ee4c90295.

I'm not exactly clear on whether you want just `check_ipc()` made non-fatal or you want all checks to be non-fatal.

Reply:
What I was *actually* wondering about is the right way to handle `sem_next_id` being unavailable in IPC namespaces, but it turns out I was just using an old kernel (yeah, I'm still using Ubuntu). https://github.com/torvalds/linux/commit/1f5c135ee509e89e0cc274333a65f73c62cb16e5 fixes this in 5.19+. I'll upgrade the kernel soon-ish and check if this really fixes the problem in my use case, but if it does, I think no changes are necessary here. Sorry for confusion.

Reply:
I see, thanks.

I think having all checks be non-fatal so `criu check` will report everything at once is more useful than the current behavior, implemented here: https://github.com/checkpoint-restore/criu/blob/18c6426eaeebc5fe7d0f9ca0acb592a3ec828b0c/criu/cr-check.c#L1434-L1511